### PR TITLE
Update title bar text color of vertical tab

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/34154

image

Resolves 

- [x] Proper title color is set after startup
- [x] Proper title color is set whenever window activation state is changed
- [x] Proper title color is set whenever toggle vertical tab or title visibility options
- [x] Proper title color is set whenever tab changes
- [x] Task manager's title color should set properly

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by simonhong** [comment]
Hmm.. it's difficult to set our colors always as frame title color is set internally from cocoa whenever related states are changed such as application theme, activation state and etc..

**Comment by simonhong** [comment]
`method_exchangeImplementations` really works nicely for browser window
but I found that it affects also non-themed dialog like below. Only themed window's title's color should be updated.
Screenshot 2023-11-16 at 2 50 54 PM

**Comment by simonhong** [comment]
Kindly ping @Brandon-T :) PTAL when you're available.

File: chromium_src/components/remote_cocoa/app_shim/native_widget_ns_window_bridge.mm [hunk 1]
```diff
@@ -52,4 +77,42 @@
...
+    return;
+  }
+
+  g_brave_title_color = skia::SkColorToDeviceNSColor(color);
+
+  // Revert previous swizzling and cache original title color.
+  // That title color will be used for non browser window frame.
+  // (ex, task manager dialog or popup window)
+  frameClassSwizzler.reset();
+  NSInvocation* invocation = [NSInvocation
```
Comment (Brandon-T) [inline chromium_src/components/remote_cocoa/app_shim/native_widget_ns_window_bridge.mm:99]: Don't use `NSInvocation` unless calling functions defined as blocks. Use messaging.
There is a much simpler and safer way to override Objective-C methods and to call methods without `NSInvocation`.

Replace this entire function with:
```ObjC
void NativeWidgetNSWindowBridge::UpdateWindowTitleColor(SkColor color) {
  NSView* frameView = window_.contentView.superview;
  if (![frameView isKindOfClass:[BrowserWindowFrame class]]) {
    return;
  }

  if (![frameView respondsToSelector:@selector(_currentTitleColor)]) {
    return;
  }

  g_brave_title_color = skia::SkColorToDeviceNSColor(color);

  // Cache original title color
  g_original_title_color = [(BrowserWindowFrame*)frameView originalTitleColor];

  // Reset title to apply new title color.
  NSString* title = window_.title;
  window_.title = @"";
  window_.title = title;
}
```

File: chromium_src/components/remote_cocoa/app_shim/native_widget_ns_window_bridge.mm [hunk 2]
```diff
@@ -3,8 +3,33 @@
...
+NSColor* __strong g_brave_title_color = nil;
+}  // namespace
+
+@interface BrowserWindowFrame : NativeWidgetMacNSWindowTitledFrame
+@end
+
+@interface FakeBrowserWindowFrame : NSObject
+@end
+
+@implementation FakeBrowserWindowFrame
```
Comment (Brandon-T) [inline chromium_src/components/remote_cocoa/app_shim/native_widget_ns_window_bridge.mm:22]: Don't swizzle, remove `FakeBrowserWindowFrame` :)

Override via Forward Declaration of `[NSThemeFrame _currentTitleColor]` and Category on `BrowserWindowFrame` instead.

```ObjC
@interface NSThemeFrame(BraveTheme)
- (NSColor*)_currentTitleColor;
@end

@interface BrowserWindowFrame : NativeWidgetMacNSWindowTitledFrame
@end

@implementation BrowserWindowFrame(BraveTheme)
- (NSColor*)_currentTitleColor {
  return g_brave_title_color;
}

- (NSColor*)originalTitleColor {
  return [super _currentTitleColor];
}
@end
```

**Review by Brandon-T — CHANGES_REQUESTED** [review]

**Review by simonhong — COMMENTED** [review]
@Brandon-T Thanks for your advice.
It's really better with category. All fixed! 👍🏼

**Review by Brandon-T — APPROVED** [review]

**Review by goodov — APPROVED** [review]

**Pull request merged**